src/main/java/no/sysco/middleware/metrics/prometheus/jdbc/JdbcCollector.java

```diff
@@ -1,77 +1,74 @@
+    private final Path configSource;
+
+    private volatile Collection<JdbcConfig> configs = List.of();
+    private volatile Instant lastUpdate = Instant.EPOCH;
+
+    private final Counter configReloadSuccess;
+    private final Counter configReloadFailure;
 
-    JdbcCollector(File in, String prefix) {
-        configFile = in;
-        metricPrefix = prefix;
+    private final Clock clock = Clock.systemUTC();
 
-        configReloadSuccess = Counter.build()
+    JdbcCollector(String metricPrefix, Path configSource) throws IOException {
+        this.configSource = requireNonNull(configSource);
+        this.metricPrefix = requireNonNull(metricPrefix);
+
+        this.configReloadSuccess = Counter.build()
                 .name(metricPrefix + "_config_reload_success_total")
                 .help("Number of times configuration have successfully been reloaded.")
                 .register();
 
-        configReloadFailure = Counter.build()
+        this.configReloadFailure = Counter.build()
                 .name(metricPrefix + "_config_reload_failure_total")
                 .help("Number of times configuration have failed to be reloaded.")
                 .register();
-        loadConfigFromFile(in);
+
+        final var lastUpdate = Files.getLastModifiedTime(configSource).toInstant();
+        loadConfig();
+        this.lastUpdate = lastUpdate;
     }
 
-    private void loadConfigFromFile(File in) {
-        Set<JdbcConfig> config = new HashSet<>();
-        lastUpdate = in.lastModified();
-        if (in.isDirectory()) {
-            for (File file : in.listFiles()) {
-                try (FileReader fr = new FileReader(file)) {
-                    config.add(new JdbcConfig((Map<String, Object>) new Yaml().load(fr)));
-                } catch (Exception iae) {
-                    LOGGER.log(Level.SEVERE, "Skipping " + file.getName() + " due to error: " + iae.getMessage(), iae);
-                }
-            }
-        } else {
-            try (FileReader fr = new FileReader(in)) {
-                config.add(new JdbcConfig((Map<String, Object>) new Yaml().load(fr)));
-            } catch (IOException e) {
-                LOGGER.log(Level.SEVERE, "Skipping " + in.getName() + " due to error: " + e.getMessage(), e);
+    private void loadConfig() throws IOException {
+        final var configs = new ArrayList<JdbcConfig>();
+        for (final var it = Files.walk(configSource).filter(Files::isRegularFile).iterator(); it.hasNext();) {
+            final var file = it.next();
+            try (final var configData = Files.newInputStream(file)) {
+                configs.add(new JdbcConfig(metricPrefix, Config.parseYaml(configData), clock));
             }
         }
 
-        if (config.size() == 0) {
-            throw new IllegalArgumentException("Error reading config files: " + in.getName());
+        if (configs.isEmpty()) {
+            throw new IllegalArgumentException("No configuration in " + configSource);
```

Comment:
I strived to retain the current behavior of the exporter here. It failed before, so I left it as is.

All in all, this method is a bit weird anyways, since it's called both from within the constructor as well as (indirectly) from the collect() method, and it ignores and logs any errors in input files. The exception type isn't ideal, but an IllegalStateException also doesn't feel entirely right.